azuredevops/internal/service/git/resource_git_repository_branch.go

```diff
@@ -0,0 +1,265 @@
+package git
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/v6/git"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/client"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/converter"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/tfhelper"
+)
+
+const (
+	REF_BRANCH_PREFIX = "refs/heads/"
+	REF_TAG_PREFIX    = "refs/tags/"
+)
+
+// ResourceGitRepositoryBranch schema to manage the lifecycle of a git repository branch
+func ResourceGitRepositoryBranch() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceGitRepositoryBranchCreate,
+		ReadContext:   resourceGitRepositoryBranchRead,
+		DeleteContext: resourceGitRepositoryBranchDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceGitRepositoryBranchImport,
+		},
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.NoZeroValues,
+			},
+			"repository_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.IsUUID,
+			},
+			"ref_branch": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ValidateFunc:  validation.StringIsNotEmpty,
+				ConflictsWith: []string{"ref_tag", "ref_commit_id"},
+			},
+			"ref_tag": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ValidateFunc:  validation.StringIsNotEmpty,
+				ConflictsWith: []string{"ref_branch", "ref_commit_id"},
+			},
+			"ref_commit_id": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				Computed:      true,
+				ValidateFunc:  validation.StringIsNotEmpty,
+				ConflictsWith: []string{"ref_branch", "ref_tag"},
+			},
+			"last_commit_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
```

Comment:
`ref_commit_id` should not be an `computed`, latest commit ID of `ref_branch` will change over development and the `ref_commit_id` will be updated as well.
My understanding of this API is `ref_branch`, `ref_commit_id` and `ref_commit_id` are conflict with each other and at least one and only one of them needs to be configured in the HCL. They need to add a `ExactlyOneOf:  []string{"ref_branch", "ref_tag", "ref_commit_id"},` constraint.

Reply:
Yea, I kind of looked at what they did in github's implementation. They only set `ref_commit_id` during creation given by user or by the other ref. It indeed creates a double meaning for `ref_commit_id`. I'll remove it for now.